flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRule.java

```diff
@@ -0,0 +1,920 @@
+            List<RexNode> projectInJoinKeyList =
+                    factJoinKeys.stream()
+                            .map(
+                                    i ->
+                                            origProgram.expandLocalRef(
+                                                    origProgram.getProjectList().get(i)))
+                            .collect(Collectors.toList());
+            List<RexNode> exprList = origProgram.getExprList();
+            for (RexNode rex : projectInJoinKeyList) {
+                if (exprList.contains(rex)
+                        && exprList.get(exprList.indexOf(rex)) instanceof RexInputRef) {
+                    joinKeysIndexInFactTable.add(exprList.indexOf(rex));
+                }
+            }
+
+            if (joinKeysIndexInFactTable.isEmpty()) {
+                return Collections.emptyList();
+            }
+
+            candidateFields =
+                    joinKeysIndexInFactTable.stream()
+                            .map(i -> factScan.getRowType().getFieldNames().get(i))
+                            .collect(Collectors.toList());
+        }
+
+        List<String> acceptedFields =
+                ((SupportsDynamicFiltering) tableSource).applyDynamicFiltering(candidateFields);
+
+        return acceptedFields.stream()
+                .map(f -> factScan.getRowType().getFieldNames().indexOf(f))
+                .collect(Collectors.toList());
+    }
+
+    private static boolean isDimSide(RelNode rel) {
+        DppDimSideFactors dimSideFactors = new DppDimSideFactors();
+        visitDimSide(rel, dimSideFactors);
+        return dimSideFactors.isDimSide();
+    }
+
+    /**
+     * Visit dim side to judge whether dim side has filter condition and whether dim side's source
+     * table scan is non partitioned scan.
+     */
+    private static void visitDimSide(RelNode rel, DppDimSideFactors dimSideFactors) {
+        if (rel instanceof TableScan) {
+            TableScan scan = (TableScan) rel;
+            TableSourceTable table = scan.getTable().unwrap(TableSourceTable.class);
+            if (table == null) {
+                return;
+            }
+            if (table.abilitySpecs() != null && table.abilitySpecs().length != 0) {
+                for (SourceAbilitySpec spec : table.abilitySpecs()) {
+                    if (spec instanceof FilterPushDownSpec) {
+                        List<RexNode> predicates = ((FilterPushDownSpec) spec).getPredicates();
+                        for (RexNode predicate : predicates) {
+                            if (isSuitableFilter(predicate)) {
+                                dimSideFactors.hasFilter = true;
+                            }
+                        }
+                    }
+                }
+            }
+
+            CatalogTable catalogTable = table.contextResolvedTable().getTable();
+            dimSideFactors.hasNonPartitionedScan = !catalogTable.isPartitioned();
```

Comment:
> if there is join which inputs are a dim table and a fact table, `hasNonPartitionedScan ` value will be false, since the fact table will be visited at last.

In this version, there is no join in dim side.